lib/proto/proto_pipe.c

```diff
@@ -162,11 +163,14 @@ callback_pipe_read(void * cookie, int status)
 	return ((P->callback)(P->cookie));
 
 eof:
-	/* We aren't going to write any more. */
-	shutdown(P->s_out, SHUT_WR);
-
-	/* Record that we have reached EOF. */
-	*(P->status) = 0;
+	/* We aren't going to write any more; ignore any ENOTSOCK. */
+	if (shutdown(P->s_out, SHUT_WR) && (errno != ENOTSOCK)) {
```

Comment:
Hmm, I'm not sure that ENOTSOCK is valid here.  These are descriptors we're calling `select()` on, so they had better be sockets.

For that matter, if we get EBADF or EINVAL those should be fatal errors -- it means something went wrong, not just that the connection is dying.  And I *think* ENOTCONN is wrong too but I'm not 100% certain.  (If a connection is being killed e.g. by a RST arriving, I think the socket still counts as being "connected" even though the TCP connection it applies to is dead?)

My inclination is to make this a `warnp` / `goto err0;` situation, possibly with special handling for ENOTCONN.